tinygrad/codegen/uopgraph.py

```diff
@@ -477,9 +481,28 @@ def delete_redundant_gates(root:UOp) -> Optional[UOp]:
   def find_gate(x:UOp) -> Optional[UOp]:
     if x.op is UOps.IF: return x
     return next((ret for s in x.src if (ret:=find_gate(s)) is not None), None)
-  if len(root.src) == 3 or (gate:=find_gate(root)) is None or gate.src[0] is not root.src[3]: return None
+  if len(root.src) == 3 or (gate:=find_gate(root)) is None or gate.src[0] is not root.src[3] \
+    and not ((sub_gate := find_gate(root.src[2])) is not None and sub_gate.src[0] is root.src[-1].src[0]):
+    return None
   return UOp(UOps.STORE, root.dtype, root.src[:3], root.arg)
 
+def merge_gates(sink:UOp) -> Optional[UOp]:
+  @functools.lru_cache(None)
+  def has_range(x:UOp) -> bool: return any(s2p.op is UOps.RANGE for s2p in x.sparents)
+  if not any(x.op is UOps.STORE and len(x.src) == 4 for x in sink.src): return None
```

Comment:
Do you mean something like `if not any(x.op is UOps.IF in sink.parents): return None`? Or do you mean checking the `x.src[-1].op is UOps.IF`?

Reply:
Sorry, I completely misunderstood your question last night.

Yes, that's clever. It should work since we shouldn't care about positioning of the IF when there's only one.